lib/cached_network_image.dart

```diff
@@ -86,81 +90,177 @@ class CacheManager {
+        _shouldStoreDataAgain = false;
+        return true;
+      }
+      _isStoringData = false;
+      return false;
+    });
+  }
+
+  _saveDataInPrefs() async{
     Map json = new Map();
+
     await synchronized(_lock, () {
-      cacheData.forEach((key, cache) {
+      _cacheData.forEach((key, cache) {
         json[key] = cache._map;
       });
     });
-    prefs.setString("lib_cached_image_data", JSON.encode(json));
+    _prefs.setString("lib_cached_image_data", JSON.encode(json));
+
+    if(await _shouldSaveAgain()){
+      await _saveDataInPrefs();
+    }
+  }
+
+  _cleanCache({force: false}) async {
+    var sinceLastClean = new DateTime.now().difference(lastCacheClean);
+
+    if (force ||
+        sinceLastClean > inbetweenCleans ||
+        _cacheData.length > maxNrOfCacheObjects) {
+      await synchronized(_lock, () async {
+        var oldestDateAllowed = new DateTime.now().subtract(maxAgeCacheObject);
+
+        //Remove old objects
+        var oldValues =
+        _cacheData.values.where((c) => c.touched.isBefore(oldestDateAllowed));
+        for (var oldValue in oldValues) {
+          await _removeFile(oldValue);
+        }
+
+        //Remove oldest objects when cache contains to many items
+        if (_cacheData.length > maxNrOfCacheObjects) {
+          var allValues = _cacheData.values.toList();
+          allValues.sort((c1, c2) => c1.touched.compareTo(c2.touched));
+          for (var i = allValues.length; i > maxNrOfCacheObjects; i--) {
+            var lastItem = allValues[i - 1];
+            await _removeFile(lastItem);
+          }
+        }
+
+        lastCacheClean = new DateTime.now();
+        _prefs.setInt("lib_cached_image_data_last_clean",
+            lastCacheClean.millisecondsSinceEpoch);
+      });
+    }
+  }
+
+  _removeFile(CacheObject cacheObject) async {
+    var file = new File(cacheObject.filePath);
+    if (await file.exists()) {
+      file.delete();
+    }
+    _cacheData.remove(cacheObject.url);
   }
 
   ///Get the file from the cache or online. Depending on availability and age
   Future<File> getFile(String url) async {
-    if (!cacheData.containsKey(url)) {
+    if (!_cacheData.containsKey(url)) {
       await synchronized(_lock, () {
-        if (!cacheData.containsKey(url)) {
-          cacheData[url] = new CacheObject();
+        if (!_cacheData.containsKey(url)) {
+          _cacheData[url] = new CacheObject(url);
         }
       });
     }
 
-    var cacheObject = cacheData[url];
+    var cacheObject = _cacheData[url];
     await synchronized(cacheObject.lock, () async {
+      // Set touched date to show that this object is being used recently
+      cacheObject.touch();
+
       //If we have never downloaded this file, do download
       if (cacheObject.filePath == null) {
-        cacheData[url] = await downloadFile(url);
+        _cacheData[url] = await downloadFile(url);
         return;
       }
       //If file is removed from the cache storage, download again
```

Comment:
I am not sure why `If we have never downloaded this file, do download` has to be different to `If file is removed from the cache storage, download again`. It can be exactly same flow:

```
 if(!exsitsInCache){
  fetchFromNetwork();
}

return fromCache;
```

Reply:
True, but you would get `new File(null).exists()`, is that allowed? Otherwise you still have to do some checks whether path == null and I don't directly see how this can be made cleaner.